Python/ceval.c

```diff
@@ -4066,7 +4066,88 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, PyFrameObject *f, int throwflag)
+            }
+        }
+        case TARGET(CALL_CFUNCTION_FAST): {
+            /* Builtin METH_FASTCALL functions, without keywords */
+            SpecializedCacheEntry *caches = GET_CACHE();
+            _PyAdaptiveEntry *cache0 = &caches[0].adaptive;
+            _PyCallCFunctionCache *cache1 = &caches[-1].call_function;
+            PyObject **callable = &PEEK(cache0->original_oparg + 1);
+            DEOPT_IF(!PyCFunction_CheckExact(*callable), CALL_FUNCTION);
+            PyCFunction cfunc = PyCFunction_GET_FUNCTION(*callable);
+            DEOPT_IF(cfunc != cache1->cfunc, CALL_FUNCTION);
+
+            assert(cframe.use_tracing == 0);
+            // res = func(self, args, nargs)
+            PyObject *res = ((_PyCFunctionFast)(void(*)(void))cfunc)(
+                PyCFunction_GET_SELF(*callable),
+                stack_pointer - cache0->original_oparg,
+                cache0->original_oparg);
+            assert((res != NULL) ^ (_PyErr_Occurred(tstate) != NULL));
+
+            /* Clear the stack of the function object. */
+            while (stack_pointer > callable) {
+                PyObject *x = EXT_POP(stack_pointer);
+                Py_DECREF(x);
+            }
+            PUSH(res);
+            if (res == NULL) {
+                /* Not deopting because this doesn't mean our optimization was wrong.
+                   `res` can be NULL for valid reasons. Eg. getattr(x, 'invalid').
+                   In those cases an exception is set, so we must handle it.
+                */
+                goto error;
+            }
+            record_cache_hit(cache0);
+            STAT_INC(CALL_FUNCTION, hit);
+            DISPATCH();
+        }
+        case TARGET(CALL_CFUNCTION_O): {
+            /* Builtin METH_O functions */
+            SpecializedCacheEntry *caches = GET_CACHE();
+            _PyAdaptiveEntry *cache0 = &caches[0].adaptive;
+            _PyCallCFunctionCache *cache1 = &caches[-1].call_function;
+            PyObject **callable = &PEEK(cache0->original_oparg + 1);
+            DEOPT_IF(!PyCFunction_CheckExact(*callable), CALL_FUNCTION);
+            PyCFunction cfunc = PyCFunction_GET_FUNCTION(*callable);
+            DEOPT_IF(cfunc != cache1->cfunc, CALL_FUNCTION);
+
+            assert(cframe.use_tracing == 0);
+            // res = func(self, arg)
+            PyObject *res = cfunc(
+                PyCFunction_GET_SELF(*callable),
+                *(stack_pointer - cache0->original_oparg));
+            assert((res != NULL) ^ (_PyErr_Occurred(tstate) != NULL));
+
+            /* Clear the stack of the function object. */
+            while (stack_pointer > callable) {
+                PyObject *x = EXT_POP(stack_pointer);
+                Py_DECREF(x);
+            }
+            PUSH(res);
+            if (res == NULL) {
+                goto error;
+            }
+            record_cache_hit(cache0);
+            STAT_INC(CALL_FUNCTION, hit);
+            DISPATCH();        
+        }
```

Comment:
I really want to refactor this and make them share code. There's only 1 line of difference between the two. But I don't like gigantic multi-line macros too :(. Are macros the only way here?